ISPN-14766 Integrate with Insights

fax4ever opened the pull request:
https://issues.redhat.com/browse/JDG-5898
https://issues.redhat.com/browse/ISPN-14766

wburns commented on core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:
```diff
@@ -370,6 +370,19 @@ public <V> ScheduledBlockingCompletableStage<V> scheduleRunBlocking(Supplier<V>
+         return null;
+      };
+
+      var scheduledStage = new ScheduledBlockingFuture<>(supplier, traceId);
```
This won't quite work as the stage will be completed on the first pass. We will need a slightly different abstraction since the task can be scheduled multiple times.

wburns replied:
I would say there is no reason for this to return a ScheduledBlockingCompletableStage as we don't ever expect to get a return value from it, unlike `scheduleRunBlocking`. In that case we can probably just return a `ScheduledFuture` like the `ScheduledExecutorService` does where get just always throws an exception but you can cancel it as needed. Then internally we can handle rescheduling the task as needed.

fax4ever replied:
you're definitively right.... fixing...

wburns commented on core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java:
```diff
@@ -370,6 +370,19 @@ public <V> ScheduledBlockingCompletableStage<V> scheduleRunBlocking(Supplier<V>
    }
 
+   @Override
+   public ScheduledBlockingCompletableStage<Void> scheduleRunBlockingAtFixedRate(Runnable runnable, long initialDelay, long period, TimeUnit unit, Object traceId) {
```
One thing to note is do we really want AtFixedRate? Is FixedDelay okay, that one is usually better as it guarantees you can't have concurrent runnables running at the same time. The drawback is the scheduling is not always done at the same offset.

fax4ever replied:
Good point. I used AtFixedRate in place of FixedDelay since the way the EAP integration works. AtFixedRate sounds the way RedHat Insights works, since we can imagine a user opens the Insights web page and expects an update at a given hour of time. Anyway I'll ask if we can switch to FixedDelay here to the Insight experts.
Unfortunately so far I haven't had the opportunity to have any contact with any Insight experts from the company, thus all I’ve done so far I just deducted looking at the Insights Java client and at EAP Insight integration :(

wburns commented on server/insights/src/main/java/org/infinispan/server/insights/InsightsModule.java:
```diff
@@ -0,0 +1,86 @@
+
+   @Override
+   public void cacheManagerStopping(GlobalComponentRegistry gcr) {
+      insightsReportController.shutdown();
```
We need a null check I believe for when insights are disabled.

wburns commented on server/insights/src/main/java/org/infinispan/server/insights/report/InfinispanSubreport.java:
```diff
@@ -0,0 +1,31 @@
+
+   @Override
+   public void generateReport() {
+      jsonReport = cacheManagerInfo.toJson().toString();
```
Seems it would be better to store the Json representation instead of the String? Then the serialization can just pass the Json representation directly instead of using raw.

fax4ever replied:
Do you mean to pass the Json representation directly to `com.fasterxml.jackson.core.JsonGenerator` instead of calling `com.fasterxml.jackson.core.JsonGenerator#writeRaw(String)`?
If so, how can I do it?

wburns replied:
We would want to add a utility to do it, but it is nothing flashy just some some nice recursive calls checking what level the json is and invoking the appropriate `write***` method on the JsonGenerator class based on the Json types we have.

e.g.

```java
Json json = cacheManagerInfo.toJson();

recursiveWrite(json, jsonGenerator);
```
```java
void recursiveWrite(Json json, JsonGenerator generator) {
   if (json.isObject()) {
      generator.startObject();
      var map = json.asJsonMap();
      for (Entry<String, Json> entry : map.entrySet()) {
         generator.writeFieldName(entry.getKey());
         recursiveWrite(entry.getValue(), generator);
      }
      generator.writeEndObject();
   } else if (
   .... rest
}
```

But I guess since this isn't a performance required operation, we can just stick with what you have.

tristantarrant replied:
I think `CacheManagerInfo` doesn't contain all the information we want. Without exposing specifics, we could use the `CacheResourceV2.CacheFullDetail` information which gives us high-level info such as `indexed`, `secure`, `transactional`, etc

tristantarrant replied:
Additionally, we should not be including "private" information in the report, such as IP addresses, hostnames, etc

wburns commented on server/insights/src/main/java/org/infinispan/server/insights/report/InfinispanTopReport.java:
```diff
@@ -0,0 +1,31 @@
+      try {
+         return super.getIdentificationName();
+      } catch (InsightsException ex) {
+         return "Infinispan Server " + cacheManagerInfo.getNodeName();
```
Do we want to completely ignore the exception?

fax4ever replied:
According to the EAP Insights integration code it seems to be safe, see https://github.com/jbossas/jboss-eap7/blob/7.4.x/insights/src/main/java/org/jboss/eap/insights/report/JBossTopLevelReport.java.
Since I haven't received any feedback from the Red Hat Insights guys I don't know if it is OK.
I just copied and adapted what has been done by the EAP team.

tristantarrant replied:
Also use `Version.brandName()`

wburns commented on server/insights/src/main/java/org/infinispan/server/insights/scheduler/InfinispanInsightsScheduler.java:
```diff
@@ -0,0 +1,89 @@
+
+   @Override
+   public List<Runnable> shutdownNow() {
+      return null;
```
```suggestion
      return List.of();
```

fax4ever replied:
Right! I'm applying it

wburns commented on server/insights/src/main/java/org/infinispan/server/insights/scheduler/InfinispanInsightsScheduler.java:
```diff
@@ -0,0 +1,89 @@
+                              "Red Hat Insights client scheduler shutdown, scheduled send failed: "
+                                    + ix.getMessage()),
+                        ix);
+                  shutdown();
```
Shutdown does nothing.. do we want to have some type of running state for the InfinispanInsightsScheduler  so we can actually stop the ScheduledFuture  as well?

fax4ever replied:
You definitely right! I agree, we need to stop the ScheduledFuture without stopping the scheduler executor.

fax4ever commented:
Thank you @wburns for the review.
I should have addressed all the points here.
I also added a commit to disable the integration from the module by default, the integration will be active only if the system property `infinispan.insights.enabled` is set to `true`.

fax4ever commented:
@tristantarrant I updated the dependency of Java Insights client to 1.0.9.
Always building it locally from the 1.0.9 tag.

tristantarrant commented on server/insights/src/main/java/org/infinispan/server/insights/InfinispanInsightsConfiguration.java:
```diff
@@ -0,0 +1,11 @@
+
+   @Override
+   public String getIdentificationName() {
+      return "Red Hat DataGrid Server";
```
Use `Version.brandName()` as that will have the correct format

tristantarrant reviewed (changes requested):
The Infinispan report needs to be changed to include higher-level info, such as cache features in uses, realm types, whether TLS is used at the endpoint and transport level, which protocols are being used, etc

fax4ever commented:
Thanks @tristantarrant I applied the changes and opened this other pull request to generate and publish the new report

fax4ever commented:
@tristantarrant I think I've made the change requested

tristantarrant commented on server/core/src/main/java/org/infinispan/server/core/BaseServerManagement.java:
```diff
@@ -0,0 +1,188 @@
+            .set("total-started", threadMXBean.getTotalStartedThreadCount());
+
+      Set<String> cacheNames = cacheManager.getCacheNames();
+      int protoSchemas = (!cacheNames.contains(PROTOBUF_METADATA_CACHE_NAME)) ? 0 :
```
```java
int protoSchemas = (!cacheNames.contains(PROTOBUF_METADATA_CACHE_NAME)) ? 0 :
            SecurityActions.getUnwrappedCache(cacheManager, PROTOBUF_METADATA_CACHE_NAME).size();
```

tristantarrant commented on server/core/src/main/java/org/infinispan/server/core/BaseServerManagement.java:
```diff
@@ -0,0 +1,188 @@
+
+      Json cpu = new BaseOperatingSystemAdditionalMetrics().cpuReport();
+
+      ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
```
I think we need a substitution for the native quarkus server.

tristantarrant commented on server/tests/src/test/java/org/infinispan/server/functional/overview/OverviewReportIT.java:
```diff
@@ -0,0 +1,68 @@
+import net.spy.memcached.ConnectionFactoryBuilder;
+import net.spy.memcached.MemcachedClient;
+
+public class OverviewReportIT {
```
I would not introduce a new IT for this, but add this to the `RESTAuthorizationTest` class so that it gets properly tested with authorization.

fax4ever replied:
The test uses REST, but also HotRod and the Memcached protocol to see the effects on the clients report.
Moving it to `RESTAuthorizationTest` I wasn't able to send a put using the Memcached client, so I had to remove the check for the `Memcached` client on the report.
I did the change in a different commit.

fax4ever replied:
It seems that it is not possible to safely merge the test into the `RESTAuthorizationTest`.
I have another idea , keep the test since we need to use HotRod and Memcached protocol from it and add a simpler form of it into `RESTAuthorizationTest` that avoids to use other protocols.

fax4ever replied:
Specifically, `OverviewReportIT` will keep to test the fact that usages of clients are reflected in the report.
While the test I'm adding to `RESTAuthorizationTest` will test the authorization behavior of the REST call. 
I don't know what you think of it.

fax4ever commented:
Made the last changes and rebased.

ryanemerson commented:
@fax4ever Can you remove the commit adding `-Pinsights` and squash the commits, then I think we're good to go.

fax4ever commented:
> @fax4ever Can you remove the commit adding `-Pinsights` and squash the commits, then I think we're good to go.

Sure @ryanemerson. Thanks for review it!
I'm rebasing now. After that I'll push the changes.
If after that CI works, I'll remove the commit adding `-Pinsights` and squash the commits.

fax4ever commented:
@ryanemerson I also added [ISPN-14766 Apply ISPN-15143 to insights module](https://github.com/infinispan/infinispan/pull/10929/commits/2811d65a0a47b4fb93f5bec7649bcd3cb4bc32d8) to make the Insight module consistent with the change of ISPN-14766.

fax4ever commented:
> @fax4ever Can you remove the commit adding `-Pinsights` and squash the commits, then I think we're good to go.

done!

tristantarrant commented:
Merged, thanks

fax4ever commented:
Thank you @tristantarrant !!!
